### DISCUSSION TITLE
Multiple provider through websocket

### DISCUSSION THREAD

**Discussion by bajrangCoder** [mkslanc/ace-linters#80] [Q&A] [answered]
Is it possible to use multiple languages provider through websocket for eg: python (through pylsp ) & svelte.

In my case , here only one Editor but dynamically I update the mode of file, how can I use here ? Is there any guide for this?

Because i have tried like this (but failed):
```js
import { AceLanguageClient } from "ace-linters/build/ace-language-client";

const lspServers =  {
   "python": "ws://localhost:3000",
   // more language provider
 };
const activeLSPConection = {};

if (activeLSPConnections[language]) {
  return this.activeLSPConnections[language];
}
const lspUrl = lspServers[language];
if (!lspUrl) {
  return;
}
let socket = new WebSocket(lspUrl);
const languageProvider = AceLanguageClient.for(socket);
activeLSPConnections[language] = socket;
languageProvider.setGlobalOptions("python", {
  pylsp: {/* configuration */},
});
languageProvider.registerEditor(editor);
window.registerFormatter("python-lsp-formatter", ["py"], () => {
  languageProvider.format();
});
```

> but this doesn't works , in console it logs only **Connection dispose** .....
and due this default snippets of ace aren't working in other modes.

**Comment by mkslanc** [comment]
For the `AceLanguageClient`, you should implement this logic on the server side, specifically where you start servers over websockets. You'll need to redirect queries based on the mode you're currently using. After sending an initialization message to all servers, you should register the document with the appropriate server (e.g., a Svelte document with a Svelte server) and unregister it when you switch modes to another server. Essentially, by the end, you'll have a setup similar to [ServiceManager](https://github.com/mkslanc/ace-linters/blob/main/packages/ace-linters/src/services/service-manager.ts#L17).

**Reply by bajrangCoder** [reply]
Ok , I will try.
But add typescript declaration (typings) of ace-linter , it will be helpful in finding the method.

**Reply by mkslanc** [reply]
Which method? Generated types [here](https://github.com/mkslanc/ace-linters/tree/main/packages/ace-linters/types)

**Reply by bajrangCoder** [reply]
Oo, sorry!
I hadn't seen this.
Got it.

**Reply by bajrangCoder** [reply]
Still it's failed.

Isn't there any way to do these stuff on frontend?
Or is there any guide to this on server as you have mentioned.

**Reply by mkslanc** [reply, marked as answer]
You could utilize [this bridge](https://github.com/mkslanc/websockets-ls) . This should make it. It could be done on frontend, but right now it's not priority for me.

**Reply by bajrangCoder** [reply]
Ok

**Reply by bajrangCoder** [reply]
Hmm, it's working fine.
And I have utilised your [bridge](https://github.com/mkslanc/websockets-ls) and tried to add lsp for php and go using ([phpactor](https://github.com/phpactor/phpactor) and [gopls](https://pkg.go.dev/golang.org/x/tools/gopls)) but both failed because they require real uri, not virtual one(session5.php) .

I guess it can be fixed, by providing real path but there is no any method for it in ace linter..

![IMG_20230911_173214_253](https://github.com/mkslanc/ace-linters/assets/71929976/023ee03c-c9de-4fa1-9d6e-36a90a5593b8)

**Reply by mkslanc** [reply]
You could easily change filename on fly in bridge - just add `__dirname` to filename, whenever you need it. You would need to add this prefix for requests going to server and remove it for request to client.

**Reply by bajrangCoder** [reply]
[here](https://github.com/mkslanc/websockets-ls/blob/91878e7b97b3d8bd06d11a9dc20de45c210f7263/server.js#L30)

?

**Reply by mkslanc** [reply]
No - everywhere where request has filename - `parsed.params.textDocument.uri`  . You could look for it  in [LSP Specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentItem) . But basically replacing `let fileName = parsed.params.textDocument.uri` with `__dirname +  fileName` should make it. Don't forget to remove `__dirname` [here ](https://github.com/mkslanc/websockets-ls/blob/91878e7b97b3d8bd06d11a9dc20de45c210f7263/server.js#L23) for client's requests

**Reply by bajrangCoder** [reply]
I guess it's not due to this path issue because I have tried with real uri then also it gives same error, In case of rust-analyzer too. I thought rust-analyzer or gopls etc is not compatible with ace-linters, as these works fine with neo-vim(binary of lsp server)

![Screenshot_20230924-133148_1](https://github.com/mkslanc/ace-linters/assets/71929976/877d3d9f-b10b-4c42-b7e1-e6b1a64d671d)
